design-proposals/feature-lifecycle.md

```diff
@@ -0,0 +1,306 @@
+
+  The API is considered stable with care no not break backward compatibility
+  with previous beta releases.
+  This implies that fields may only be added during this stage,
+  not removed or renamed.
+
+  The feature functionality is controlled using a Feature-Gate (FG) during
+  runtime. It must be specified for the feature to be active.
+
+- **GA**:
+  The feature graduated to general-availability (GA) and is now part of
+  the core features.
+
+  The feature functionality is no longer controlled by a FG.
+
+> **Warning**: A Feature Gate flag is solely intended to control
+> feature lifecycle. It should not be confused and used as a cluster
+> configurable enablement of the functionality.
+> In cases where the cluster admin should control a functionality,
+> regardless to the feature stage, dedicated configuration field/s
+> should be included.
+
+#### Removal
+When and if a feature is targeted for deprecation and retirement,
+it needs to pass a deprecation process, depending on its current
+release stage (Alpha, Beta, GA).
+
+For more details, see [here](#deprecation-and-removal).
+
+#### Release Stage Transition Table
+The following table summarized the different release stages with their
+transition requirements and restrictions.
+
+| Stage          | Period range    | F.Gate | Removal Availability       |
+|----------------|-----------------|--------|----------------------------|
+| Alpha          | 1 to 2 releases | YES    | Between **minor** releases |
+| Beta           | 1 to 3 releases | YES    | Between **minor** releases |
+| GA             | -               | NO     | Between **major** releases |
+
+Through Alpha and Beta feature releases, a FG must be set in order
+for the feature to function.
+By default, no FG is specified, therefore the feature is by disabled.
+
+If a feature is not able to transition to the next stage in the defined period,
+it should be removed automatically.
+
+> **Note**: Exceptions to the period range may apply
+> if 2/3 of active maintainers come to agreement to prolong
+> a specific feature.
+
+### Deprecation and Removal
+One reason for features to go through the Alpha and Beta stages,
+is the opportunity to examine their usefulness and adoption.
+Same goes with major releases that intentionally allow breaking
+backward compatibility (as specified by [semver](https://semver.org/)).
+
+Therefore, it is only natural that some features will not graduate
+between the stages, or will be found irrelevant after some time and be
+removed when transitioning between major releases.
+
+#### Major Releases
+KubeVirt follows semver versioning, in which major versions may
+break API compatibility. Therefore, discontinuation of features
+is somehow simpler when incrementing the major version.
+
+However, this is not without a cost.
+When a new major release is introduced, the previous one is still maintained
+and supported, something that does not exist with minor releases.
```

Comment:
This is expected, but not guaranteed - that two or more major versions are supported at the same time.

Reply:
Per the K8S documentations I reviewed, it is required for at least two major versions to be supported in order to allow proper upgrade. I guess Kubevirt may consider not to support at least two, but the implications are not fully clear to me at this stage.

For clarification, upgrades between major releases usually require support for rollback, mainly due to the backward comparability issues.

Reply:
Note: we do not support rollback

Reply:
Yes, I was wrong - I don't know what was on my mind!